### DISCUSSION TITLE
can't enter in ITEM_COMMAND

### DISCUSSION THREAD

**Discussion by Sebidarkprince** [forntoh/LcdMenu#120] [Q&A]
Hi dear, 
I'm actually using your LcdMenu.h library but I have problem to enter in the ITEM_COMMAND.
I am beginner in programming, could you help me?
When I enter to the ITEM_COMMAND to call the function "remplissage40g()" nothing happen

Thank you for your help :)

`/*
 Basic Menu

 https://lcdmenu.forntoh.dev/examples/basic

*/
#include <LiquidCrystal_I2C.h>
#include <LcdMenu.h>
#include <ItemSubMenu.h>
#include <ItemCommand.h>
#include "AiEsp32RotaryEncoder.h"
#include "Arduino.h"
#include "HX711.h"

#define DEBUG_HX711

// calibration parameter from calibrating code with known values
#define CALIBRATION_FACTOR 210
 
#define RPWM 19 //  RPWM BTS7960
#define LPWM 18 //  LPWM BTS7960
#define PWM 23//   pin R_EN and L_EN BTS7960

#define LCD_ROWS 4
#define LCD_COLS 20

int a=0;
// data pin and clock pin HX711
byte pinData = 16;
byte pinClk = 4;

// define HX711
HX711 scale;

 //Configure keyboard keys (ASCII)
#define UP 56        // NUMPAD 8
#define DOWN 50      // NUMPAD 2
#define LEFT 52      // NUMPAD 4
#define RIGHT 54     // NUMPAD 6
#define ENTER 53     // NUMPAD 5
#define BACK 55      // NUMPAD 7
#define BACKSPACE 8  // BACKSPACE
#define CLEAR 46     // NUMPAD .

extern MenuItem* menuprincipalMenu[];
extern MenuItem* remplissageMenu[];
extern MenuItem* manuelMenu[];
extern MenuItem* reglagesMenu[];
extern MenuItem* doseaMenu[];
extern MenuItem* dosebMenu[];
extern MenuItem* dosecMenu[];
extern MenuItem* dosedMenu[];
extern MenuItem* doseeMenu[];

#define ROTARY_ENCODER_A_PIN 33 //DT
#define ROTARY_ENCODER_B_PIN 32 //CLK
#define ROTARY_ENCODER_BUTTON_PIN 27
#define ROTARY_ENCODER_VCC_PIN -1
#define ROTARY_ENCODER_STEPS 4
AiEsp32RotaryEncoder rotaryEncoder = AiEsp32RotaryEncoder(
    ROTARY_ENCODER_A_PIN, 
    ROTARY_ENCODER_B_PIN, 
    ROTARY_ENCODER_BUTTON_PIN, 
    ROTARY_ENCODER_VCC_PIN
    );
  /*LCD*/
// Create the lcd object address 0x3F and 16 columns x 2 rows 
LiquidCrystal_I2C lcd (0x27, LCD_ROWS,LCD_COLS);

void motor_CW() {
  digitalWrite (LPWM, LOW);
  digitalWrite (RPWM, HIGH);
  analogWrite (PWM,500);
  Serial.println ("CW");
}
void motor_CCW() {
  digitalWrite (LPWM, HIGH);
  digitalWrite (RPWM, LOW);
  analogWrite (PWM,500); 
  Serial.println ("CCW");
}
void motor_stop() {
  digitalWrite (LPWM, LOW);
  digitalWrite (RPWM, LOW);
  analogWrite (PWM,0);
  Serial.println ("STOP");
}

void remplissage40g() {
  Serial.println ("test");
while(scale.get_units()<40)
{
motor_CW();
}
motor_stop();
  }

MAIN_MENU(

//Page d'accueil
ITEM_SUBMENU("Abeilles&Papillons", menuprincipalMenu)
);
  //Menu principal
  SUB_MENU(menuprincipalMenu, mainMenu,
  ITEM_SUBMENU("Remplissage", remplissageMenu),
  ITEM_SUBMENU("Manuel", manuelMenu),
  ITEM_SUBMENU("Reglages", reglagesMenu)
  );

    //Menu remplissage
    SUB_MENU(remplissageMenu, menuprincipalMenu,
    ITEM_SUBMENU("40 g", doseaMenu),
    ITEM_SUBMENU("125 g", dosebMenu),
    ITEM_SUBMENU("250 g", dosecMenu),
    ITEM_SUBMENU("500 g", dosedMenu),
    ITEM_SUBMENU("1 kg", doseeMenu),
    ITEM_SUBMENU("Retour", menuprincipalMenu)
    );

      //Menu 40g
      SUB_MENU(doseaMenu, remplissageMenu,
      ITEM_COMMAND("Remplissage", remplissage40g),
      ITEM_SUBMENU("Retour", remplissageMenu)
      );
      //Menu 125g
      SUB_MENU(dosebMenu, remplissageMenu,
      ITEM_BASIC("oui"),
      ITEM_BASIC("non"),
      ITEM_SUBMENU("Retour", remplissageMenu)
      );
      //Menu 250g
      SUB_MENU(dosecMenu, remplissageMenu,
      ITEM_BASIC("oui"),
      ITEM_BASIC("non"),
      ITEM_SUBMENU("Retour", remplissageMenu)
      );
      //Menu 500g
      SUB_MENU(dosedMenu, remplissageMenu,
      ITEM_BASIC("oui"),
      ITEM_BASIC("non"),
      ITEM_SUBMENU("Retour", remplissageMenu)
      );
      //Menu 1kg
      SUB_MENU(doseeMenu, remplissageMenu,
      ITEM_BASIC("oui"),
      ITEM_BASIC("non"),
      ITEM_SUBMENU("Retour", remplissageMenu)
      );

    //Menu manuel
    SUB_MENU(manuelMenu, menuprincipalMenu,
    ITEM_BASIC("Marche avant"),
    ITEM_BASIC("Marche arrière"),
    ITEM_SUBMENU("Retour", menuprincipalMenu)
    );

    //Menu régalges
    SUB_MENU(reglagesMenu, menuprincipalMenu,
    ITEM_BASIC("Stop-goutte"),
    ITEM_BASIC("Infos"),
    ITEM_SUBMENU("Retour", menuprincipalMenu)
    );

LcdMenu menu(LCD_ROWS, LCD_COLS);

void rotary_onButtonClick()
{
    static unsigned long lastTimePressed = 0; // Soft debouncing
    if (millis() - lastTimePressed < 500)
    {
            return;
    }
    lastTimePressed = millis();
    Serial.print("button pressed ");
    Serial.print(millis());
    Serial.println(" milliseconds after restart");
}

void rotary_loop() {
	//first lets handle rotary encoder button click
	if (rotaryEncoder.currentButtonState() == BUT_RELEASED) {
		//we can process it here or call separate function like:
    	rotary_onButtonClick();
	}
 if (rotaryEncoder.isEncoderButtonClicked())
    {
         Serial.print("Clicked");  
         menu.enter();
    }
	//lets see if anything changed
	int8_t encoderDelta = rotaryEncoder.encoderChanged();
	
	//optionally we can ignore whenever there is no change
	if (encoderDelta == 0) return;
	
    //for some cases we only want to know if value is 
    //increased or decreased (typically for menu items)
	if (encoderDelta>2) {Serial.print("+");  menu.down();}
  
	if (encoderDelta<-2) {Serial.print("-");  menu.up();}

  if (rotaryEncoder.currentButtonState() == BUT_PUSHED){Serial.print("Clicked");  menu.enter();}

    //for other cases we want to know what is current value. 
    //Additionally often we only want if something changed
	//example: when using rotary encoder to set termostat temperature, or sound volume etc
	
	//if value is changed compared to our last read
	if (encoderDelta!=0) {
		//now we need current value
		int16_t encoderValue = rotaryEncoder.readEncoder();

		//process new value. Here is simple output.
		Serial.print("Value: ");
		Serial.println(encoderValue);

	}
}

void setup() {
  rotaryEncoder.begin();
  rotaryEncoder.setup([]{rotaryEncoder.readEncoder_ISR();});
    Serial.begin(9600);
    // Initialize LcdMenu with the menu items
    menu.setupLcdWithMenu(0x27, mainMenu);

//Initializing sensor
  scale.begin(pinData, pinClk);
  // apply the calibration value
  scale.set_scale(CALIBRATION_FACTOR);
  // Initialize the tare
  //Assuming there is no weight on the scale at start up, reset the scale to 0
  scale.tare();

  pinMode (RPWM, OUTPUT);
  pinMode (PWM, OUTPUT);
  pinMode (LPWM, OUTPUT);
}

void loop() {
  
  

  rotary_loop();
    if (!Serial.available()) return;
    char command = Serial.read();

  
  if (command == UP)
        menu.up();
    else if (command == DOWN)
        menu.down();
    else if (command == LEFT)
        menu.left();
    else if (command == RIGHT)
        menu.right();
    else if (command == ENTER)
        menu.enter();
    else if (command == BACK)
        menu.back();
}`

**Comment by forntoh** [comment]
Could you please better format the code. It's not easy to understand it.
